File: Unit 2 - Graphics/src/main/java/com/gyoge/itcsgraphics/drawables/Snowflake.java
package com.gyoge.itcsgraphics.drawables;

import java.util.Random;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Polygon;

/**
 * A simple snowflake.
 *
 * @author Yogesh Thambidurai
 * @see com.gyoge.itcsgraphics.animators.SnowflakeAnimator
 */
public class Snowflake implements Drawable {

    private static final Random random = new Random();
    private double x;
    private double y;

    public Snowflake(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return this.x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return this.y;
    }

    public void setY(double y) {
        this.y = y;
    }

    @SuppressWarnings("DuplicatedCode")
    public void draw(Graphics goon) {
        int intX = (int) this.x;
        int intY = (int) this.y;
        int color = random.nextInt(55) + 200;
        goon.setColor(new Color(color, color, color));

        goon.drawLine(intX - 3, intY, intX + 3, intY);
        goon.drawLine(intX, intY - 3, intX, intY + 3);
        goon.drawLine(intX - 3, intY - 3, intX + 3, intY + 3);
        goon.drawLine(intX + 3, intY - 3, intX - 3, intY + 3);

        goon.drawPolygon(new Polygon(
            new int[]{intX - 2, intX + 2, intX, intX, intX - 2, intX + 2, intX - 2, intX - 2},
            new int[]{intY, intY, intY - 2, intY + 2, intY - 2, intY + 2, intY + 2, intY - 2},
            8
        ));
    }

}

File: Unit 2 - Graphics/src/main/java/com/gyoge/itcsgraphics/drawables/Snowflake.kt
package com.gyoge.itcsgraphics.drawables

import java.awt.Color
import java.awt.Graphics
import java.awt.Polygon
import java.util.*

/**
 * A simple snowflake.
 *
 * @author Yogesh Thambidurai
 * @see com.gyoge.itcsgraphics.animators.SnowflakeAnimator
 * @see Drawable
 */
class Snowflake(x: Int, y: Int) : Drawable {
    var x: Double
    var y: Double

    /**
     * Constructs a new snowflake.
     */
    init {
        this.x = x.toDouble()
        this.y = y.toDouble()
    }

    override fun draw(goon: Graphics) {
        // Values to sub in
        val x: Int = this.x.toInt()
        val y: Int = this.y.toInt()

        val left: Int = x - 3
        val right: Int = x + 3
        val up: Int = y + 3
        val down: Int = y - 3

        // Random white hue
        val color: Int = random.nextInt(55) + 200
        goon.color = Color(color, color, color)

        // Diagonals
        goon.drawLine(left, y, right, y)
        goon.drawLine(x, up, x, down)
        goon.drawLine(left, up, right, down)
        goon.drawLine(right, up, left, down)

        // Center bit
        goon.drawPolygon(
            Polygon(
                intArrayOf(x - 1, x + 1, x, x, x - 1, x + 1, x - 1, x + 1),
                intArrayOf(y, y, y - 1, y + 1, y - 1, y + 1, y + 1, y - 1),
                8
            )
        )
    }

    companion object {
        /** Random object to use. */
        private val random = Random()
    }

}